connection/connection.go

```diff
@@ -93,149 +121,262 @@ type DBConnection struct {
+	return row, nil
 }
 
-// Query executs `Query` with transaction.
-func (c *TxConnection) Query(ctx context.Context, conn *sql.DB, query string, args ...interface{}) (*sql.Rows, error) {
+// Query executes `Query` with transaction.
+func (c *TxConnection) Query(ctx context.Context, conn Connection, query string, args ...interface{}) (*sql.Rows, error) {
 	if err := c.beginIfNotInitialized(conn); err != nil {
 		return nil, errors.WithStack(err)
 	}
-	if ctx == nil {
-		rows, err := c.tx.Query(query, args...)
-		if err != nil {
-			return nil, errors.WithStack(err)
+	tx := c.dsnToTx[conn.DSN()]
+	rows, err := func() (*sql.Rows, error) {
+		if ctx == nil {
+			return tx.Query(query, args...)
 		}
-		return rows, nil
-	}
-	rows, err := c.tx.QueryContext(ctx, query, args...)
+		return tx.QueryContext(ctx, query, args...)
+	}()
 	if err != nil {
 		return nil, errors.WithStack(err)
 	}
+	c.ReadQueries = append(c.ReadQueries, &QueryLog{
+		Query: query,
+		Args:  args,
+	})
 	return rows, nil
 }
 
-// Exec executs `Exec` with transaction.
-func (c *TxConnection) Exec(ctx context.Context, conn *sql.DB, query string, args ...interface{}) (sql.Result, error) {
+// Exec executes `Exec` with transaction.
+func (c *TxConnection) Exec(ctx context.Context, conn Connection, query string, args ...interface{}) (sql.Result, error) {
 	if err := c.beginIfNotInitialized(conn); err != nil {
 		return nil, errors.WithStack(err)
 	}
-	if ctx == nil {
-		result, err := c.tx.Exec(query, args...)
-		if err != nil {
-			return nil, errors.WithStack(err)
+	tx := c.dsnToTx[conn.DSN()]
+	result, err := func() (sql.Result, error) {
+		if ctx == nil {
+			return tx.Exec(query, args...)
 		}
-		return result, nil
+		return tx.ExecContext(ctx, query, args...)
+	}()
+	if err != nil {
+		return nil, errors.WithStack(err)
 	}
-	result, err := c.tx.ExecContext(ctx, query, args...)
+	id, err := result.LastInsertId()
 	if err != nil {
 		return nil, errors.WithStack(err)
 	}
+	queryLog := &QueryLog{
+		Query:        query,
+		Args:         args,
+		LastInsertID: id,
+	}
+	c.txToWriteQueries[tx] = append(c.txToWriteQueries[tx], queryLog)
+	c.WriteQueries = append(c.WriteQueries, queryLog)
 	return result, nil
 }
 
-// Commit executs `Commit` with transaction.
-func (c *TxConnection) Commit() error {
+// Commit executes `Commit` with transaction.
+func (c *TxConnection) Commit() (e error) {
 	if c == nil {
-		return errors.New("cannot commit. TxConnection is nil")
+		return nil
+	}
+	if len(c.dsnToTx) == 0 {
+		return nil
+	}
+	if err := c.BeforeCommitCallback(); err != nil {
+		return errors.WithStack(err)
+	}
+	committedWriteQueryNum := 0
+	failedWriteQueries := []*QueryLog{}
+	isCriticalError := false
+
+	defer func() {
+		if len(failedWriteQueries) == 0 {
+			if err := c.AfterCommitSuccessCallback(); err != nil {
+				e = err
+			}
+		} else if len(failedWriteQueries) > 0 {
+			if err := c.AfterCommitFailureCallback(isCriticalError, failedWriteQueries); err != nil {
+				e = err
```

Comment:
I considered to use `errors.WithStack(err)` instead of `err`

What does this look like?

Reply:
Added stack trace for `e` . Thanks !